### PR TITLE
[aggregation/simple,component,core] introduce BucketStrategy

### PR DESCRIPTION
**Pull request by udoprog**
This moves the specific implementation for how a bucket strategy is performed into the `BucketStrategy` interface.

This is intended to set us up so that we can eventually switch from the current `(end - extent, end]` (`BucketStrategy.END`) bucket interval, to `[start, start + extent)` (`BucketStrategy.START`).

To maintain backwards compatibility when the alternative implementation is introduced, a new feature (`com.spotify.heroic.end_bucket_strategy`) is introduced and enabled by default.

### CODE REVIEW

File: heroic-core/src/main/java/com/spotify/heroic/metric/LocalMetricManager.java [hunk 1]
```diff
@@ -163,17 +165,22 @@ public boolean isEmpty() {
...
             final QueryContext queryContext = request.getContext();
+            final Features features = request.getFeatures();
 
             queryLogger.logIncomingRequestAtNode(queryContext, request);
 
-            final boolean slicedFetch = request.getFeatures().hasFeature(Feature.SLICED_DATA_FETCH);
+            final boolean slicedFetch = features.hasFeature(Feature.SLICED_DATA_FETCH);
+
+            final BucketStrategy bucketStrategy =
+                options.getBucketStrategy().orElse(BucketStrategy.END);
```
Comment (gabrielgerhardsson) [inline heroic-core/src/main/java/com/spotify/heroic/metric/LocalMetricManager.java]: I believe you want to use features.hasFeature(Feature.END_BUCKET) here?

Comment (udoprog) [inline heroic-core/src/main/java/com/spotify/heroic/metric/LocalMetricManager.java]: There's no other bucket strategy right now. The new strategy and that check will be introduced in a separate PR.

Comment (gabrielgerhardsson) [inline heroic-core/src/main/java/com/spotify/heroic/metric/LocalMetricManager.java]: Yes I understand. It's just that Feature.END_BUCKET is unused as the code is right now. I would have expected it to be hooked up, just for completeness of the commit, even if the !hasFeature path would be just an exception.
No biggie tho of course :)

Comment (udoprog) [inline heroic-core/src/main/java/com/spotify/heroic/metric/LocalMetricManager.java]: It has been changed.
